contracts/utils/cryptography/draft-ERC7739Utils.sol

```diff
@@ -0,0 +1,288 @@
+        bytes memory contentsType
+    ) internal pure returns (bytes memory) {
+        return abi.encodePacked(signature, separator, contents, contentsType, uint16(contentsType.length));
+    }
+
+    /**
+     * @dev Nests a `contents` digest into an {EIP712} type that simulates the `eth_personalSign` RPC
+     * method in the context of smart contracts.
+     *
+     * This typed uses the {_NESTED_PERSONAL_SIGN_TYPEHASH} type to nest the `contents` hash for the
+     * current domain `separator`.
+     *
+     * To produce a signature for this nested type, the signer must sign the following type hash:
+     *
+     * ```solidity
+     * bytes32 hash = keccak256(abi.encodePacked(
+     *      \x19\x01,
+     *      CURRENT_DOMAIN_SEPARATOR,
+     *      keccak256(
+     *          abi.encode(
+     *              keccak256("PersonalSign(bytes prefixed)"),
+     *              keccak256(abi.encode("\x19Ethereum Signed Message:\n32",contents))
+     *          )
+     *      )
+     * ));
+     * ```
+     */
+    function toNestedPersonalSignHash(bytes32 separator, bytes32 contents) internal pure returns (bytes32) {
+        return
+            MessageHashUtils.toTypedDataHash(
+                separator,
+                keccak256(abi.encode(_NESTED_PERSONAL_SIGN_TYPEHASH, MessageHashUtils.toEthSignedMessageHash(contents)))
+            );
+    }
+
+    /**
+     * @dev Nests an {EIP712} typed data `contents` digest into an {EIP712} that simulates
+     * `eth_signTypedData` RPC method for another domain (application).
+     *
+     * This type uses the {NESTED_TYPED_DATA_TYPEHASH} type to nest the `contents` hash
+     * for the current domain with the application's domain separator.
+     *
+     * To produce a signature for this nested type, the signer must sign the following type hash:
+     *
+     * ```solidity
+     * bytes32 hash = keccak256(
+     *  abi.encodePacked(
+     *      \x19\x01,
+     *      separator, // The domain separator of the application contract
+     *      keccak256(
+     *          abi.encode(
+     *              NESTED_TYPED_DATA_TYPEHASH(contentsType), // See {NESTED_TYPED_DATA_TYPEHASH}
+     *              contents,
+     *              // See {IERC5267-eip712Domain} for the following arguments from the verifying contract's domain
+     *              keccak256(bytes(name)),
+     *              keccak256(bytes(version)),
+     *              chainId,
+     *              verifyingContract,
+     *              salt,
+     *              keccak256(abi.encodePacked(extensions))
+     *          )
+     *      )
+     *  )
+     *);
+     *
+     * NOTE: The arguments should be those of the verifying application. See {EIP712-_domainSeparatorV4}
+     * and {IERC5267-eip712Domain} for more details of how to obtain these values. Respectively, they
+     * must be obtained from the verifying contract (e.g. an Account) and the application domain.
+     */
+    function toNestedTypedDataHash(
```

Comment:
Do we really need that function ? Why not use `MessageHashUtils.toTypedDataHash` directly ?

Reply:
Good point. It's result of multiple rounds of iteration but I decided to keep it just for the instructions. I'd leave it for the completeness of this library. It'd be fine if you guys decide to remove it.

Reply:
@frangio wdyt ?

Reply:
Slight preference for keeping this function. But I also was wondering whether it made sense, so I'd be ok with removing it.